SI-2458 Import shadows def in other unit

som-snytt opened the pull request:
Fixes the implementation of point 4. from SLS chapter 2:

> Definitions made available by a package clause not in the
> compilation unit where the reference occurs have lowest precedence.

We had code to deal with this:

if (defSym.exists && impSym.exists) {
if (isPackageOwnedInDifferentUnit(defSym))
defSym = NoSymbol

But, in the `pos/t2458`, no `impSym` was around, due to filtering
in `depthOk`. The depth of the imported symbol was the same as the
depth of the defined symbol, which didn't pass:

imp.depth > symbolDepth

`depthOk` already had special handling for Java sources: if the
import is non-wildcard and depths are equal, retain the import.

This commit does the same thing for Scala sources, using the
`isPackageOwnedInDifferentUnit` as the criterion. It also makes
the check for the "imported ... is permanently hidden" warning
to avoid the spurious warning.

Fix by retronym
Excavation and preparation of bones for mounting by community

Fixes scala/bug#2458
Fixes scala/bug#9552

som-snytt committed:
SI-2458 Import shadows def in other unit

Fixes the implementation of point 4. from SLS chapter 2:

> Definitions made available by a package clause not in the
> compilation unit where the reference occurs have lowest precedence.

We had code to deal with this:

if (defSym.exists && impSym.exists) {
if (isPackageOwnedInDifferentUnit(defSym))
defSym = NoSymbol

But, in the `pos/t2458`, no `impSym` was around, due to filtering
in `depthOk`. The depth of the imported symbol was the same as the
depth of the defined symbol, which didn't pass:

imp.depth > symbolDepth

`depthOk` already had special handling for Java sources: if the
import is non-wildcard and depths are equal, retain the import.

This commit does the same thing for Scala sources, using the
`isPackageOwnedInDifferentUnit` as the criterion. It also makes
the check for the "imported ... is permanently hidden" warning
to avoid the spurious warning.

Fix by retronym
Excavation and preparation of bones for mounting by community

som-snytt commented:
This doesn't fix SI-9552. The import should be higher precedence than definition in other file. Maybe the package object fools a depth check?
```
package object p {
  type Foo = Int 
} 
```
and separate unit
```
package p {
  object Dingo { type Foo = String }
  import Dingo._
  trait Bippy  {
    // import Dingo._  // this version compiles
    def z: Foo
    z: String
  }
}
```

som-snytt committed:
SI-9552 Wildcard import precedence vs foreign definition

The Java test for explicit import tripped up the Scala
test for foreign definition (i.e., in other compilation
unit). This commit refactors the boolean expression.

som-snytt commented:
Fixes the Java check for explicit import. Scala lets wildcard import hide foreign definition.

som-snytt commented:
/rebuild 83fb60c

som-snytt commented:
/rebuild

som-snytt commented:
Suffers from the broken build.

som-snytt commented:
Some documentation notes the bug and should be updated:

http://docs.scala-lang.org/tutorials/FAQ/finding-implicits

som-snytt commented:
/rebuild

Blaisorblade commented:
@som-snytt what happened to this PR? Are the bugtracker entries up-to-date currently? (https://issues.scala-lang.org/browse/SI-2458 is closed, https://issues.scala-lang.org/browse/SI-9552 is open, and I'm confused).

som-snytt commented:
@Blaisorblade The old issue was closed prematurely when the spec was clarified; this PR was meant to be implementation which also nails down the newer issue (which also objects to the behavior as specified). I don't remember the build status, so I'll have to revisit it.

Blaisorblade commented:
@som-snytt Thanks — it sounds like all the open parts are tracked by the newer issue. I dread unfixed closed issues and forgotten bugs.

som-snytt commented:
/rebuild

som-snytt commented:
Previous failure was due to cycle in `ProcessBuilder.scala` with wildcard import from companion.

There was also an import in scaladoc that was not to spec.

som-snytt commented:
This idiom now inoculates against accidental local dir named `util`:
```
package p

import scala._
import util.Try

class C {
  def f = Try(42)
}
```
Too bad that's not more useful.

som-snytt commented on build.sbt:
```diff
@@ -102,6 +102,9 @@ lazy val commonSettings = clearSourceAndResourceDirectories ++ publishSettings +
   // Avoid circular dependencies for scalaInstance (see https://github.com/sbt/sbt/issues/1872)
   managedScalaInstance := false,
+  //maxErrors := 500,
+  maxErrors := 5,
```
Meant to set that to zero.

som-snytt commented on src/compiler/scala/reflect/reify/Reifier.scala:
```diff
@@ -69,7 +70,7 @@ abstract class Reifier extends States
 
           val tpe = typer.packedType(tree, NoSymbol)
-          val ReifiedType(_, _, tpeSymtab, _, rtpe, tpeReificationIsConcrete) = `package`.reifyType(global)(typer, universe, mirror, tpe, concrete = false)
+          val ReifiedType(_, _, tpeSymtab, _, rtpe, tpeReificationIsConcrete) = reifyTypeGlobal(global)(typer, universe, mirror, tpe, concrete = false)
```
This was good, when you got packages and methods named reify.

som-snytt commented on src/compiler/scala/tools/nsc/reporters/EditReporter.scala:
```diff
@@ -0,0 +1,39 @@
+/**
+ * This class implements a Reporter that stores erroneous files in a script for editing.
+ */
+class EditReporter(settings: Settings) extends Reporter {
```
If @lrytz was up for a reporters module, I'd contribute.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -984,6 +984,14 @@ trait Contexts { self: Analyzer =>
+    // For that case, check if the root import dealiases to the definition.
+    private def resolveAmbiguity(name: Name, defSym: Symbol, impSym: Symbol, imp: ImportInfo): Boolean = {
+      defSym == impSym || imp.isRootImport && {
+        impSym.isAliasType && impSym.info.typeSymbol.fullName == defSym.fullName
```
@retronym Help. I didn't have time to improve this yet.

retronym replied:
I'm having trouble paging in the context for this from my mental swap space... could you elaborate a little?

som-snytt replied:
If root `import scala._` alias `Seq` collides with `Seq` defined in enclosing package, recognize that they are the same. There is related code for when two imports collide.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -1109,37 +1112,47 @@ trait Contexts { self: Analyzer =>
+        else if (depthCheckEnabled) foreignDefined
+        else imp.depth == symbolDepth && foreignDefined
       )
+      def depthCheckEnabled = true  // settings.isScala213
```
I was going to add the kill switch just for the last commit, where package-defined sym is amby with import in enclosing scope, the cause of many subtle edits, but I'd have to move those edits to 2.13.x.

retronym commented:
Heads up for https://github.com/scala/scala/pull/6297 which is likely to collide with this.

som-snytt commented:
Don't forget to update the spec deficiency about root imports and aliasing. https://gitter.im/scala/contributors?at=5a7b33b96117191e610894a1

SethTisue commented:
rebase?

som-snytt commented:
What really sucks about this PR is that moving it to 2.13 collides with collections realignment.

My plan is to re-do this on 2.13 and then add `-Yimport` to make preamble imports tractable. Not sure exactly how it will work, but similarly to the leading `import Predef._` exclusion, a source file can opt out of the preamble; however, it seems to me that the syntax should align with ordinary binding.

If I compile with `-Yimport:my.Predef._` then `import my.Predef.{_ => _}` turns it off with no special magic.

If I compile with `-Yimport:my.Predef.any2stringadd`, then `import my.Predef.{_ => _}` does not hide? Unless the rule is construed that "level 4" includes the preamble imports. I don't think anyone considered that option, i.e., the preamble is not just stuck at the front of your source; it has level 4 "binding force" like bindings from a package clause that is not in the current unit.

som-snytt committed:
Preamble imports are lowest priority binding

This change allows the package-owned symbols clause
to respect nesting depth without breaking all
code. This corresponds to the intuition that
the implicit imports of the root context, which
are not written in source code, do not conflict
with user-written imports and definitions.

som-snytt commented:
Rebased with the fix for root imports aka preamble imports. No longer impacts the world like an asteroid.

som-snytt committed:
Fix imports in two tests

This has the form
```
import p._ ; package p { S }
```
where S is both made available in p and also explicitly
imported in enclosing scope.

The code for reconciling ambiguous imports is not
triggered here.

som-snytt commented:
/rebuild

som-snytt commented:
Gitter is down? so I'll ask here whether this seems too loose:

```
scala> class C { type InputStream = java.io.InputStream }
defined class C

scala> def f(c: C) = { import java.io._, c._ ; null.asInstanceOf[InputStream] }
f: (c: C)c.InputStream

scala> def f(c: C) = { import c._, java.io._ ; null.asInstanceOf[InputStream] }
f: (c: C)java.io.InputStream

```

som-snytt committed:
Yimports for preamble imports

Packages and objects can be specified.

A source file can exclude imports from
one or more objects with the usual mechanism,
formerly known as the `Predef` exclusion clause.

Handle overloaded foreign definitions:
If one alternative is foreign definition,
take the overload as foreign definition.

SethTisue commented:
/sync

som-snytt commented:
I needed this today.

```
imported `Configurator' is permanently hidden by definition of object Configurator in package apps
```

som-snytt commented:
I need this basically all the time.
```
[warn] /Users/andrew/workspace/scala/test/benchmarks/src/main/scala/scala/tools/reporters/ReportingBenchmark.scala:9: imported `Reporter' is permanently hidden by definition of class Reporter in package reporters
```

adriaanm commented:
Moving this to the next 2.13 milestone.

som-snytt commented:
@adriaanm There's a forward port PR for 2.13; I have a `-Xsource:2.13` commit to push here; would you consider that?

som-snytt commented:
Superseded for 2.13 by https://github.com/scala/scala/pull/6589
